conans/client/installer.py

```diff
@@ -467,18 +468,38 @@ def _build_package(self, node, output, keep_build, remotes):
         return pref
 
     @staticmethod
-    def _propagate_info(node):
+    def _propagate_info(node, using_build_profile):
         # Get deps_cpp_info from upstream nodes
         node_order = [n for n in node.public_closure if n.binary != BINARY_SKIP]
         # List sort is stable, will keep the original order of the closure, but prioritize levels
         conan_file = node.conanfile
-        transitive = node.transitive_closure.values()
+        transitive = [it for it in node.transitive_closure.values()]
+
+        br_host = []
+        for it in node.dependencies:
+            if it.require.build_require_context == CONTEXT_HOST:
+                br_host.extend(it.dst.transitive_closure.values())
+
         for n in node_order:
             if n not in transitive:
                 conan_file.output.info("Applying build-requirement: %s" % str(n.ref))
-            conan_file.deps_cpp_info.update(n.conanfile.cpp_info, n.ref.name)
-            conan_file.deps_env_info.update(n.conanfile.env_info, n.ref.name)
-            conan_file.deps_user_info[n.ref.name] = n.conanfile.user_info
+
+            if not using_build_profile:  # Do not touch anything
+                conan_file.deps_user_info[n.ref.name] = n.conanfile.user_info
+                conan_file.deps_cpp_info.update(n.conanfile.cpp_info, n.ref.name)
+                conan_file.deps_env_info.update(n.conanfile.env_info, n.ref.name)
+            else:
+                if n in transitive or n in br_host:
+                    conan_file.deps_cpp_info.update(n.conanfile.cpp_info, n.ref.name)
+                else:
+                    env_info = EnvInfo()
+                    env_info._values_ = n.conanfile.env_info._values_.copy()
+                    # Add cpp_info.bin_paths/lib_paths to env_info (it is needed for runtime)
+                    env_info.DYLD_LIBRARY_PATH.extend(n.conanfile.cpp_info.lib_paths)
+                    env_info.DYLD_LIBRARY_PATH.extend(n.conanfile.cpp_info.framework_paths)
+                    env_info.LD_LIBRARY_PATH.extend(n.conanfile.cpp_info.lib_paths)
+                    env_info.PATH.extend(n.conanfile.cpp_info.bin_paths)
+                    conan_file.deps_env_info.update(env_info, n.ref.name)
```

Comment:
@memsharded I think that this can lead to problems :disappointed: 

The information in `deps_env_info` is added to the environment with the `get_env_context_manager(conanfile)` that runs before entering the `conanfile::build()` function (and the rest of functions). It will add many libraries to the search path that could interfere with the applications that run inside the `build()` function...

----

The solution would be not to populate the environment before entering the `build()` but inside the `self.run("....", run_environment=True)`, so these paths will be added on demand and would affect only to that command.

Reply:
...but relying on that behavior (previous comment) would make it impossible to add a BR from a profile if the recipe is not prepared: when using the system one it needs `run_environment=False`, when using the BR one it needs `run_environment=True`.